### ISSUE TITLE
tvOS support?

### ISSUE THREAD

**Issue by appsird** [52inc/Pulley#322]
Hello,

Can Pulley be used in tvOS applications?

thanks.

**Comment by brendan09** [comment]
I’m honestly not sure. I’ve never tried it. If you decide to try it, I’d definitely be interested in knowing if it works.

**Comment by appsird** [comment]
There appears integration issues via Podfile

You can try as follows. 

[1] First Create a new tvOS project in Version 10.1 (10B61)
File->New->Project->select tvOS, then Single View App. Use Product name Test
[2] Next, attempt to integrate via Podfile using 

`source 'https://github.com/CocoaPods/Specs.git'

use_frameworks!

target 'Test' do
  platform :tvos, '12.0'
  pod 'Pulley', :git => 'https://github.com/52inc/Pulley.git', :branch => 'swift-4.2'
end
`
The peform:

pod update
pod install

I then receive error messages:

Analyzing dependencies
Pre-downloading: `Pulley` from `https://github.com/52inc/Pulley.git`, branch `swift-4.2`
[!] The platform of the target `Test` (tvOS 12.0) is not compatible with `Pulley (2.5.0)`, which does not support `tvos`.

___

so it appears somehow you need to compile with tvOS support. Would appreciate this capability.

**Comment by brendan09** [comment]
Yes, that error is on purpose. We don’t mark Pulley as supporting tvOS as an officially supported platform so cocoapods gives you an error if you try to use it in a tvOS app.

If you’d like to try Pulley in a tvOS app, you can copy the Pulley source files into your project and compile it that way (assuming tvOS doesn’t have an issue with the code). 

However, we don’t have plans to make tvOS an officially supported platform.

**Comment by appsird** [comment]
>If you decide to try it, I’d definitely be interested in knowing if it works.

Ok, too close out. It does not currently working using CocoaPods.

**Comment by brendan09** [comment]
Just to point out: The cocoapods limitation is artificial. It’s a config setting. You’re welcome to try a manual import of the files.

**Comment by appsird** [comment]
I will try a manual import and report back later.

**Comment by appsird** [comment]
I invited you to a first cut of tvOS supported Pulley. 

I created a single view tvOS project. Then, installed the Pulley files. Created a separate target for the Pulley framework you should compile first. Then compile the Test app. 

Note there's 1 error stating Pulley cannot be located on an import. Would appreciate understanding why.

Once it compiles, I have included your original iOS storyboard. tvOS has a single 4K screen size. So show home I need to map your storyboard onto this storyboard, and else should 'just work'.  On this link:

https://forums.developer.apple.com/thread/95890

waynehead states one can copy the XML directly. 

Once compiled, you should be able to select appleTV simulator 4K and run. Hoping we can work together to make this work, as it appears it should. You can then use the code for a new branch of the official pulley.

**Comment by brendan09** [comment]
Because Pulley is no longer in a separate framework you don’t need to import it. If it were still in the cocoapod you would need the import. You can delete that line.

You can follow the instructions in the README for setting up a new storyboard to work with Pulley (or not use storyboards at all). It should be pretty straightforward. You don’t need to use the one provided at all.

My current employer doesn’t allow me to make code contributions, but I’ve left this project in the hands of @ulmentflam for maintenance. You can invite him and he may be interested in exploring this with you.

**Comment by appsird** [comment]
ok, will do. I commented out the three occurrences of import Pulley, and a new set of tvOS errors results which I need to navigate.

<img width="406" alt="Screen Shot 2019-03-15 at 11 59 51 AM" src="https://user-images.githubusercontent.com/8471833/54455861-00397c80-471a-11e9-86e0-e5480ea63b80.png">

**Comment by brendan09** [comment]
Yes, those APIs that Pulley uses are unavailable on tvOS. 

You’ll need to remove references to them or do an available() condition on them so they’re only includes on iOS builds.

**Comment by appsird** [comment]
Some progress in latest. Unfortunately two fairly important methods appear unavailable for override in tvOS.

<img width="1731" alt="Screen Shot 2019-03-15 at 1 19 43 PM" src="https://user-images.githubusercontent.com/8471833/54459891-36c8c480-4725-11e9-8e5e-d9781ae4caed.png">

**Comment by brendan09** [comment]
For your purposes you can use @available to mark them iOS only, or delete them in your local copy if just you are using this fork. 

TvOS doesn’t have a status bar so these aren’t needed there.

**Comment by appsird** [comment]
I used below to remove code from tvOS compile. 

```code

    #if os(iOS)
    override open var childForStatusBarStyle: UIViewController? {
        get {
            if drawerPosition == .open {
                return drawerContentViewController
            }
            
            return primaryContentViewController
        }
    }
    #endif
    
    #if os(iOS)
    override open var childForStatusBarHidden: UIViewController? {
        get {
            if drawerPosition == .open {
                return drawerContentViewController
            }
            
            return primaryContentViewController
        }
    }
    #endif
```

This removes the code, though oddly I receive a linker error. States Pulley framework not found, though it is there.

<img width="1731" alt="Screen Shot 2019-03-15 at 9 53 38 PM" src="https://user-images.githubusercontent.com/8471833/54470863-d65d7580-476c-11e9-800a-3672c99ab07d.png">

**Comment by brendan09** [comment]
That’s because it’s trying to link pulley as a module. Make sure it’s not still in your podfile (and then pod install) and make sure nothing is using the line import Pulley anywhere.

**Comment by appsird** [comment]
ok, it now compiles / runs - pops a blank screen using the default tvOS storyboard. Next will attempt to swap in Pulley Storyboard

**Comment by appsird** [comment]
swap in results in error message regarding incompatible iOS storyboard. Will need to generate a Storyboard within project which mimics the original. will attempt this now.

<img width="1593" alt="Screen Shot 2019-03-15 at 10 33 24 PM" src="https://user-images.githubusercontent.com/8471833/54471201-87b2da00-4772-11e9-953b-693d5bbb9b30.png">

**Comment by brendan09** [comment]
The storyboard in the pulley sample project is just a sample. You should make a new storyboard either way. 

Instructions on using Pulley with storyboards are in the README.

**Comment by appsird** [comment]
I was able to bring up the primary view. Though, lacking assistance from your proxy collaborator, my efforts seem at a standstill. It appears it's possible to port to tvOS, though it will take some further effort, as you use programmatic constraints, and the drawer needs further setup. If @ulmentflam accepts, I'd be happy to work with him on this.

**Comment by ulmentflam** [comment]
@appsird Sorry for the delay in responding to you, I am significantly tied up in the amount of time I have over the next few months. If there is something specific you would like me to help with, I would be happy to take a look at it when I have the time. Just create an issue in the repo you invited me to and I will take a look at it.
